language-plutus-core/src/Language/PlutusCore/Normalize.hs

```diff
@@ -1,237 +1,60 @@
-normalizeTypeM builtin@TyBuiltin{}           =
-    pure $ NormalizedType builtin
-
-{- Note [Normalizing substitution]
-@substituteNormalize[M]@ is only ever used as normalizing substitution that receives two already
-normalized types. However we do not enforce this in the type signature, because
-1) it's perfectly correct for the last argument to be non-normalized
-2) it would be annoying to wrap 'Type's into 'NormalizedType's
--}
-
--- See Note [Normalizing substitution].
--- | Substitute a type for a variable in a type and normalize in the 'NormalizeTypeM' monad.
-substNormalizeTypeM
-    :: (HasUnique (tyname ann) TypeUnique, MonadQuote m)
-    => NormalizedType tyname ann                                -- ^ @ty@
-    -> tyname ann                                               -- ^ @name@
-    -> Type tyname ann                                          -- ^ @body@
-    -> NormalizeTypeT m tyname ann (NormalizedType tyname ann)  -- ^ @NORM ([ty / name] body)@
-substNormalizeTypeM ty name = withExtendedTypeVarEnv name ty . normalizeTypeM
+import           Control.Monad                          ((>=>))
 
 -- See Note [Normalization].
 -- | Normalize a 'Type'.
 normalizeType
     :: (HasUnique (tyname ann) TypeUnique, MonadQuote m)
-    => m () -> Type tyname ann -> m (NormalizedType tyname ann)
-normalizeType countStep = runNormalizeTypeM countStep . normalizeTypeM
+    => m () -> Type tyname ann -> m (Normalized (Type tyname ann))
+normalizeType countStep = rename >=> runNormalizeTypeM countStep . normalizeTypeM
 
 -- See Note [Normalization].
 -- | Normalize a 'Type' without dealing with gas.
 normalizeTypeDown
     :: (HasUnique (tyname ann) TypeUnique, MonadQuote m)
-    => Type tyname ann -> m (NormalizedType tyname ann)
+    => Type tyname ann -> m (Normalized (Type tyname ann))
 normalizeTypeDown = normalizeType $ pure ()
 
+-- | Normalize a 'Type' in the gas-consuming way.
+-- Count a single substitution step by subtracting @1@ from available gas or
+-- fail when there is no available gas.
+normalizeTypeGas
+    :: (HasUnique (tyname ann) TypeUnique, MonadQuote m)
+    => Gas -> Type tyname ann -> m (Maybe (Normalized (Type tyname ann)))
+normalizeTypeGas gas = rename >=> runNormalizeTypeGasM gas . normalizeTypeM
+
 -- | Normalize every 'Type' in a 'Term'.
 normalizeTypesIn
-    :: (HasUnique (tyname ann) TypeUnique, MonadQuote m)
-    => Term tyname name ann -> NormalizeTypeT m tyname ann (Term tyname name ann)
-normalizeTypesIn = go where
-    -- | Normalize a 'Type' and return the result as a 'Type' (as opposed to 'NormalizedType').
-    normalizeReturnType = fmap getNormalizedType . normalizeTypeM
+    :: (HasUnique (tyname ann) TypeUnique, HasUnique (name ann) TermUnique, MonadQuote m)
+    => m () -> Term tyname name ann -> m (Term tyname name ann)
+normalizeTypesIn countStep = rename >=> runNormalizeTypeM countStep . normalizeTypesInM
+
+-- | Normalize every 'Type' in a 'Term' without dealing with gas.
+normalizeTypesDownIn
```

Comment:
I'm sure we discussed this before, but I still find the "Down" suffix mystifying when I read it. Maybe `complete`?

Reply:
I think we talked about a different, even more confusing suffix (maybe we talked about that one too, but then I forgot the discussion). `Complete` is a bit longish, maybe `Full`?

Reply:
`Full` seems okay. Troll suggestion: `Omega`?

Reply:
Done.

Reply:
> Troll suggestion: `Omega`?

Nice one :)